### DISCUSSION TITLE
Using SparkLinker on large dataset

### DISCUSSION THREAD

**Discussion by shivam221098** [moj-analytical-services/splink#2012] [Q&A] [answered]
When I use SparkLinker on 1 million rows with 8 comparison fields and most of the field has 2 and 3 comparison level and one field has 5 comparison level.

When I estimate `u` probability with DuckDB, it takes 20-30 seconds to run but when I use use it with spark it takes almost 4-5 minutes, which is fine for me. But when I scale it to 5 million rows, just estimating `u` probability takes hours, I'm not able to understand what's the bottleneck for the scaling.

My Spark configuration is 
* 1 master node and 4 worker node
* All master and worker has 16GBs of RAM with Debian 12, 500GBs secondary storage.

Do I need to create more worker nodes?

**Comment by jericksonclinicaloptions** [comment]
You might look at the spark metrics and logs to see if there isn't a memory or CPU bottleneck. On Databricks I use a 4 worker node cluster but with more memory for 3 million records.

Note the primary driver of run time is the blocking rules: https://moj-analytical-services.github.io/splink/topic_guides/performance/drivers_of_performance.html. I was having performance issues until I adjusted the blocking.

Of course, you don't want to train on your entire dataset, usually you would train on a smaller dataset and run your entire dataset using a performant model you have chosen.

**Reply by shivam221098** [reply]
Thanks, @jericksonclinicaloptions, I saw I had some blocking issues on 5M records when I was doing predictions. I solved that.

I appreciate your help.

**Comment by RobinL** [comment]
The runtime for the estimate u routine is determined by the [ max_pairs ](https://github.com/moj-analytical-services/splink/blob/8b865c3cf0e4678d35ac7b0c46dd85ddcc8ced6d/splink/linker.py#L1422) parameter, not the size of the input data so the results you're seeing are surprising. 

If you could share your code, I might be able to suggest the source of the issue

**Reply by shivam221098** [reply]
@RobinL You're right. `max_pairs` param is the issue with the runtime. when I use `1e7`, It works fine. As per docs I see the greater the param this is, the more will be accuracy, and as my input has 1 million rows, I was using `1e10` as `max_pairs`, for this much large input data, I assume I have to use more than `1e10`? But that will consume more time to execute.

Here is my code that's I'm using

```python
#!/usr/bin/env python
# coding: utf-8

# In[1]:


import pandas as pd
import numpy as np
import re
import os
import matplotlib.pyplot as plt
import seaborn as sns
import splink.spark.comparison_library as cl
import splink.spark.comparison_template_library as ctl
from unidecode import unidecode
from splink.charts import waterfall_chart
from datetime import datetime
from pyspark import SparkContext, SparkConf
from pyspark.sql import SparkSession
from pyspark.sql import types
from splink.datasets import splink_datasets
from splink.spark.linker import SparkLinker
from splink.spark.blocking_rule_library import block_on
from splink.spark.jar_location import similarity_jar_location
from utils_spark import first_name_comparision, location_comparison


# ### Create Spark Session

# In[4]:


conf = SparkConf()


# In[5]:


conf.set("spark.driver.memory", "14g")
conf.set("spark.default.parallelism", "256")
conf.set("spark.executor.memory", "12g")
conf.set("spark.executor.cores", "4")
conf.set("spark.executor.instances", "16")
conf.setAppName("spark-app")


# In[6]:


path = similarity_jar_location()
conf.set("spark.jars", path)


# In[7]:


sc = SparkContext.getOrCreate(conf=conf)


# In[8]:


spark = SparkSession(sc)
spark.sparkContext.setCheckpointDir("gs://dataproc-staging/notebooks/jupyter/tmp_checkpoints")


# In[9]:


spark

# In[10]:


df = pd.read_csv("sample_fake.csv", low_memory=False)


# In[11]:


blocking_rules = list(map(block_on, [
    "block_on"
]))
blocking_rules


# In[12]:


comparisions = [
    cl.exact_match("lastname"),
    first_name_comparision("firstname", damerau_levenshtein_thresholds=[], jaro_winkler_thresholds=[], term_frequency_adjustments=True),
    first_name_comparision("secondname", damerau_levenshtein_thresholds=[], jaro_winkler_thresholds=[], term_frequency_adjustments=True),
    ctl.email_comparison("email", include_username_fuzzy_level=False, include_username_match_level=True, include_domain_match_level=True),
    cl.exact_match("unique_identifier"),
    location_comparison("company_name", "postal_code", "city", "state", "country", term_frequency_adjustments=True, include_term_frequency_adjustment_in_cols=["company_name"])
]
comparisions


# In[13]:


settings = {
    "link_type": "dedupe_only",
    "comparisons": comparisions,
    "blocking_rules_to_generate_predictions": blocking_rules,
    "retain_matching_columns": True,
    "retain_intermediate_calculation_columns": True
}

spark_df = spark.createDataFrame(df)
linker = SparkLinker(spark_df, settings, spark=spark, repartition_after_blocking=True)


# In[14]:


linker


# In[15]:


# probability of two random records match
deterministic_rules = [
    "l.block_on = r.block_on"
]

linker.estimate_probability_two_random_records_match(deterministic_rules, recall=0.8)


# In[16]:


linker.count_num_comparisons_from_blocking_rule(block_on("block_on"))


# In[ ]:


# The larger the random sample, the more accurate the predictions.
linker.estimate_u_using_random_sampling(max_pairs=1e10)  # u estimation (acc. to docs)


# In[ ]:


linker.estimate_m_from_label_column("unique_identifier")

```

My utils_spark.py file
[utils_spark.py.zip](https://github.com/moj-analytical-services/splink/files/14481879/utils_spark.py.zip)

My Spark configurations are:
* 1 Master & 4 Worker nodes
* All nodes have 2 Cores CPU, 16 GB memory, 500 GB standard persistent disk (n2-standard-4) on GCP

**Reply by shivam221098** [reply]
A quick workaround that I found to make it work is. 

* Estimate `m` and `u` probabilities from `DuckDB`. Save the model.
* Estimate `m` and `u` probabilities from `Spark` on very small input data, suppose 1K. Save the model
* We actually can't load a `DuckDB` model into `SparkLinker`. So I copied `m` and `u` values from DuckDB model to Spark model. 
* Loaded the model into `SparkLinker` and predictions worked.

```python
import json
import srsly

with open("models/linker_model_spark_temp.json") as file:
    new_model_data = json.loads(file.read())

with open("models/linker_model_duckdb.json") as file:
    old_model_data = json.loads(file.read())

for odt, ndt in zip(old_model_data['comparisons'], new_model_data['comparisons']):
    for ocl, ncl in zip(odt['comparison_levels'], ndt['comparison_levels']):
        # pop out sql_condition key from old data
        ocl.pop("sql_condition")
        ncl.update(ocl)

new_model_data['probability_two_random_records_match'] = old_model_data['probability_two_random_records_match']

srsly.write_json("models/linker_model_spark.json", new_model_data)
```

**Reply by RobinL** [reply, marked as answer]
A value of 1e9 (1 billion comparison) should be sufficient in almost all cases (in fact, in most cases even 1e7 or 1e8 should be fine).  1e10 is higher than most people usually use, and probably higheer than necessary

**Reply by shivam221098** [reply]
Thanks @RobinL.

I will try with lower values.

Thanks for your help.
